evadb/functions/llms/openai.py

```diff
@@ -0,0 +1,113 @@
+# coding=utf-8
+# Copyright 2018-2023 EvaDB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import os
+from typing import List
+
+from retry import retry
+
+from evadb.functions.llms.base import BaseLLM
+from evadb.utils.generic_utils import (
+    try_to_import_openai,
+    try_to_import_tiktoken,
+    validate_kwargs,
+)
+
+_DEFAULT_MODEL = "gpt-3.5-turbo"
+_DEFAULT_PARAMS = {
+    "model": _DEFAULT_MODEL,
+    "temperature": 0.0,
+    "request_timeout": 30,
+    "max_tokens": 1000,
+}
+
+
+class OpenAILLM(BaseLLM):
+    @property
+    def name(self) -> str:
+        return "OpenAILLM"
+
+    def setup(
+        self,
+        openai_api_key="",
+        **kwargs,
+    ) -> None:
+        super().setup(**kwargs)
+
+        try_to_import_openai()
+        import openai
+
+        openai.api_key = self.openai_api_key
+        if len(openai.api_key) == 0:
+            openai.api_key = os.environ.get("OPENAI_API_KEY", "")
+        assert (
+            len(openai.api_key) != 0
+        ), "Please set your OpenAI API key using SET OPENAI_API_KEY = 'sk-' or environment variable (OPENAI_API_KEY)"
+
+        validate_kwargs(kwargs, allowed_keys=_DEFAULT_PARAMS.keys())
+        self.model_params = {**_DEFAULT_PARAMS, **kwargs}
+
+        self.model_name = self.model_params["model"]
+
+    def generate(self, prompts: List[str]) -> List[str]:
+        import openai
+
+        @retry(tries=6, delay=20)
+        def completion_with_backoff(**kwargs):
+            return openai.ChatCompletion.create(**kwargs)
+
+        results = []
+
+        for prompt in prompts:
+            def_sys_prompt_message = {
+                "role": "system",
+                "content": "You are a helpful assistant that accomplishes user tasks.",
+            }
+
+            self.model_params["messages"].append(def_sys_prompt_message)
+            self.model_params["messages"].extend(
+                [
+                    {
+                        "role": "user",
+                        "content": f"Complete the following task: {prompt}",
+                    },
+                ],
+            )
+
+            response = completion_with_backoff(**self.model_params)
+            answer = response.choices[0].message.content
+            results.append(answer)
+
+        return results
+
+    def get_cost(self, prompt: str, response: str):
+        try_to_import_tiktoken()
+        import tiktoken
+
+        encoding = tiktoken.encoding_for_model(self.model_name)
```

Comment:
If we already have the response, we can directly compute the cost using the `response["usage"]` parameter?
Tiktoken would be good to estimate the cost before executing the query (helpful for query opt). Maybe we can have two functions, estimate_cost and get_cost. Estimating cost is not simple, though, because we do not know the completion tokens apriori. We would then need a heuristic for the estimated completion tokens.

Reply:
Yes, a valid concern. I was also thinking about it.